### PR TITLE
[IMP]hr_recruitment: Back2Basic - Improve Recruitment usability

### PR DESCRIPTION
**Pull request by FrancoisGe**
Task: https://www.odoo.com/web#id=2284842&action=333&active_id=1251&model=project.task&view_type=form&cids=1&menu_id=4720
# [IMP] hr_recruitment: Back2Basic - Improve Recruitment usability
### Form view: 
- Using a notebook view for increased clarity
Kanban view:
- Added a toggle button to publish/unpublish a job offer.
- Added buttons to directly share link trackers for Facebook, LinkedIn and Twitter.
- In the menu, added the « Create Interview Form » button to create an interview form.

### Generale: 
- Use of the "Applicant: Acknowledgement" template by default when an application is moved to the "First Interview" stage.
Form view:
- Possibility to consult an interview after this one and not before
- The "CREATE EMPLOYEE" button no longer directly creates an employee. The employee will be created if the "SAVE" button is pressed.

### Adding fields: 
- nb_hired = number of people hired from this source
- nb_application = number of people who applied from this source.
- cost = the cost used for this source
- campaign_id = campaign related to this source. By default, utm_campaign_job.
- link_tracker_id = link tracker to track the source url.
- clicks: number of clicks on the url.
### Tree view:
- Added a button to access link tracker statistics

## Hr.employee
###Generale:
- Added of two new certifications allowing to better match with the degrees present in a hr.applicant.

## Other
- Added a new tour for the module hr_recruitment.
- And some other minor style and name changes.

### CODE REVIEW

File: addons/hr_recruitment/static/src/js/tours/hr_recruitment.js [hunk 1]
```diff
@@ -6,36 +6,97 @@ var tour = require('web_tour.tour');
...
+}, {
+    trigger: ".oe_kanban_card",
+    extra_trigger: '.o_kanban_applicant',
+    content: _t("<b>Click to view</b> the application."),
+    position: "bottom"
+}, {
+    trigger: ".o_Chatter .o_ChatterTopbar_buttonSendMessage",
+    extra_trigger: '.o_applicant_form',
+    content: _t("<div><b>Try to send an email</b> to the applicant.</div><div><i>Tips: All emails sent or received are saved in the history here</i>"),
+    position: "bottom"
```
Comment (seb-odoo) [inline addons/hr_recruitment/static/src/js/tours/hr_recruitment.js:78]: I believe the issue is that you are showing the marker "too quickly", but I cannot test because the app doesn't even install on runbot.

From what I understand from your video, the marker is currently positioned on the "send message" button where it would be from the right-border of the screen if the chatter was positioned on the right, except that due to the screen size being too small, the chatter, that is first rendered as-if on the right, is then moved to the bottom, where it takes the full width of the screen, which makes "send message" move but the marker is already set in the middle.

Related file is `mail_enterprise/static/src/widgets/form_renderer/form_renderer.js`. The file is full of hacks that we can't get rid of until the form view is converted to OWL. And `mail/static/src/widgets/form_renderer/form_renderer.js`.

As a first step, I would advise to test on community first, where the chatter is always at the bottom, and see how that works there. Then depending on the result, we'll see what to do.

File: addons/hr_recruitment/static/src/js/tours/hr_recruitment.js [hunk 2]
```diff
@@ -6,36 +6,97 @@ var tour = require('web_tour.tour');
...
+}, {
```
Comment (seb-odoo) [inline addons/hr_recruitment/static/src/js/tours/hr_recruitment.js:79]: Shouldn't we have an extra step here asking the user to write on the textarea?

**Comment by tivisse** [comment]
@FrancoisGe Could you please add the TaskID on your commit messages ? Thanks :)

File: addons/hr_recruitment/static/src/js/tours/hr_recruitment.js (hunk 2 above)
Comment (FrancoisGe) [reply addons/hr_recruitment/static/src/js/tours/hr_recruitment.js:79]: No, we've already discussed with FLL about this

Comment (seb-odoo) [reply addons/hr_recruitment/static/src/js/tours/hr_recruitment.js:79]: Ok, but then what is the reasoning for not doing it? Because it seems pretty counter-intuitive to me to onboard someone by telling him to click on a button that doesn't work.

File: addons/hr_recruitment/static/src/js/tours/hr_recruitment.js (hunk 1 above)
Comment (FrancoisGe) [reply addons/hr_recruitment/static/src/js/tours/hr_recruitment.js:78]: I just tested on the Community version and I have the same bug.

There is a runbot with the right version of the code if you want to test.

Comment (seb-odoo) [reply addons/hr_recruitment/static/src/js/tours/hr_recruitment.js:78]: Ok thanks for checking it, I'll have a look on runbot.

File: addons/hr_recruitment/static/src/js/tours/hr_recruitment.js (hunk 2 above)
Comment (fll-odoo) [reply addons/hr_recruitment/static/src/js/tours/hr_recruitment.js:79]: Hello,
2 Reasons :
First one, we try to have as steps as possible
Secondly, when we click on the button "Send messages" our cursor is already in the message box where we have "Write something". So, if we add a step, it seems like a duplicate

File: addons/hr_recruitment/models/hr_recruitment.py [hunk 3]
```diff
@@ -454,33 +528,33 @@ def create_employee_from_applicant(self):
                 applicant.partner_id = new_partner_id
                 address_id = new_partner_id.address_get(['contact'])['contact']
             if applicant.partner_name or contact_name:
-                employee = self.env['hr.employee'].create({
-                    'name': applicant.partner_name or contact_name,
-                    'job_id': applicant.job_id.id or False,
-                    'job_title': applicant.job_id.name,
+                if applicant.type_id:
+                    certificate = 'other'
+                    for degree in ['graduate', 'bachelor', 'master', 'doctor']:
```
Comment (tivisse) [inline addons/hr_recruitment/models/hr_recruitment.py:533]: This simply doesn't work. 

What if we don't work in English ? To remove.

File: addons/hr_recruitment/models/hr_recruitment.py [hunk 4]
```diff
@@ -454,33 +528,33 @@ def create_employee_from_applicant(self):
...
+                        if applicant.type_id and degree in applicant.type_id.name.lower():
+                            certificate = degree
+                else:
+                    certificate = False
+                employee_data = {
+                    'default_name': applicant.partner_name or contact_name,
+                    'default_job_id': applicant.job_id.id or False,
```
Comment (tivisse) [inline addons/hr_recruitment/models/hr_recruitment.py:540]: The fallback is already False, no ``or`` needed.

File: addons/hr_recruitment/models/hr_recruitment.py [hunk 5]
```diff
@@ -454,33 +528,33 @@ def create_employee_from_applicant(self):
...
-                    applicant.job_id.message_post(
-                        body=_('New Employee %s Hired', applicant.partner_name if applicant.partner_name else applicant.name),
-                        subtype_xmlid="hr_recruitment.mt_job_applicant_hired")
-                applicant.message_post_with_view(
-                    'hr_recruitment.applicant_hired_template',
-                    values={'applicant': applicant},
-                    subtype_id=self.env.ref("hr_recruitment.mt_applicant_hired").id)
-
+                    'default_work_phone': applicant.department_id and applicant.department_id.company_id
+                            and applicant.department_id.company_id.phone or False,
```
Comment (tivisse) [inline addons/hr_recruitment/models/hr_recruitment.py:549]: Equivalent to 
```suggestion
                    'default_work_phone': applicant.department_id.company_id.phone,
```

File: addons/hr_recruitment/views/hr_job_views.xml [hunk 6]
```diff
@@ -65,14 +65,17 @@
                                         </div>
                                     </div>
                                 </t>
-                                <div name="kanban_boxes" class="row o_recruitment_kanban_boxes">
+                                
+                                <div name="kanban_boxes" class="row o_recruitment_kanban_boxes ">
```
Comment (tivisse) [inline addons/hr_recruitment/views/hr_job_views.xml:69]: This space doesn't seem needed right ?

File: addons/hr_recruitment/views/hr_job_views.xml [hunk 7]
```diff
@@ -65,14 +65,17 @@
...
                                     <div class="o_recruitment_kanban_box o_kanban_primary_bottom bottom_block" style="padding-left:8px;">
                                         <div class="col-6"></div>
-                                        <div class="col-6">
-                                            <a type="action" name="%(action_hr_job_sources)d">Link Trackers</a>
+                                        <div class="col-6 o_link_trackers">
+                                            <a role="button" name="%(hr_recruitment.action_hr_job_sources)d" type="action" class="btn btn-sm ">
+                                                <span title='Link Trackers'><i class='fa fa-lg fa-envelope' role="img" aria-label="Link Trackers"/></span>
```
Comment (tivisse) [inline addons/hr_recruitment/views/hr_job_views.xml:74]: Use double quote please ;)

File: addons/hr_recruitment/views/hr_recruitment_views.xml [hunk 8]
```diff
@@ -152,7 +154,7 @@
             <div class="oe_chatter">
                 <field name="message_follower_ids"/>
                 <field name="activity_ids"/>
-                <field name="message_ids" options="{'post_refresh': 'recipients', 'open_attachments': True}"/>
```
Comment (tivisse) [inline addons/hr_recruitment/views/hr_recruitment_views.xml:155]: Why do you remove this option ?

File: addons/hr_recruitment/views/hr_recruitment_views.xml [hunk 9]
```diff
@@ -286,14 +300,14 @@
                     <t t-name="kanban-box">
                         <div t-attf-class="{{!selection_mode ? 'oe_kanban_color_' + kanban_getcolor(record.color.raw_value) : ''}} oe_kanban_card oe_kanban_global_click oe_applicant_kanban oe_semantic_html_override">
                             <div class="o_dropdown_kanban dropdown">
-                                <a class="dropdown-toggle o-no-caret btn" role="button" data-toggle="dropdown" href="#" aria-label="Dropdown menu" title="Dropdown menu">
+                                <a class="dropdown-toggle o-no-caret btn" role="button" data-toggle="dropdown" href="#" aria-label="Dropdown menu" title="Dropdown menu" data-display="static">
```
Comment (tivisse) [inline addons/hr_recruitment/views/hr_recruitment_views.xml:303]: What does it do ? Being curious here.

File: addons/hr_recruitment/views/hr_recruitment_views.xml [hunk 10]
```diff
@@ -415,21 +429,21 @@
         <field name="model">hr.job</field>
         <field name="priority">200</field>
         <field name="arch" type="xml">
-            <form string="Create a Job Position">
+            <form string="Create a Job Position" class="o_hr_job_simple_form" >
                 <sheet>
                     <group>
-                        <field name="name" class="oe_inline" placeholder="e.g. Sales Manager"/>
+                        <field name="name" class="o_job_name oe_inline" placeholder="e.g. Sales Manager"/>
```
Comment (tivisse) [inline addons/hr_recruitment/views/hr_recruitment_views.xml:435]: This class doesn't seem to be used.

File: addons/hr_recruitment/views/hr_recruitment_views.xml [hunk 11]
```diff
@@ -415,21 +429,21 @@
...
                         <label for="alias_name" string="Application email" attrs="{'invisible': [('alias_domain', '=', False)]}" help="Define a specific contact address for this job position. If you keep it empty, the default email address will be used which is in human resources settings"/>
                         <div name="alias_def" attrs="{'invisible': [('alias_domain', '=', False)]}">
                             <field name="alias_id" class="oe_read_only" string="Email Alias" required="0"/>
                             <div class="oe_edit_only" name="edit_alias">
-                                <field name="alias_name" class="oe_inline"/>@<field name="alias_domain" class="oe_inline" readonly="1"/>
+                                <field name="alias_name" class="oe_inline o_job_alias" placeholder="e.g. sales-manager"/>@<field name="alias_domain" class="oe_inline" readonly="1"/>
```
Comment (tivisse) [inline addons/hr_recruitment/views/hr_recruitment_views.xml:440]: This one too. They look like a remaining dead development 

Could you please take a look at the whole PR and remove the dead classes ?

File: addons/hr_recruitment/views/hr_recruitment_views.xml [hunk 12]
```diff
@@ -857,14 +871,19 @@ action = act
...
+            <tree string="Sources of Applicants" editable="top" class="o_recruitment_list" sample="1">
+                <field name="source_id" placeholder="e.g. LinkedIn" decoration-bf="1" attrs="{'readonly': [('id', '!=', False)]}"/>
+                <field name="campaign_id" optional="hide" attrs="{'readonly': [('id', '!=', False)]}"/>
+                <field name="job_id" attrs="{'readonly': [('id', '!=', False)]}"/>
+                <field name="currency_id" invisible="1"/>
+                <field name="cost" widget="monetary" sum="Total Cost"/>
+                <field name="nb_hired"/>
+                <field name="nb_application"/>
+                <field name="email" attrs="{'invisible': [('email', '=', False)]}" widget="email"/>
+                <button name="create_alias" string="GENERATE EMAIL" class="btn btn-primary" type="object" attrs="{'invisible': [('email', '!=', False)]}"/>
```
Comment (tivisse) [inline addons/hr_recruitment/views/hr_recruitment_views.xml:883]:
```suggestion
                <button name="create_alias" string="Generate Email" class="btn btn-primary" type="object" attrs="{'invisible': [('email', '!=', False)]}"/>
```

File: addons/hr_recruitment/views/hr_recruitment_views.xml [hunk 13]
```diff
@@ -857,14 +871,19 @@ action = act
...
             </tree>
         </field>
-    </record>
+    </record>
```
Comment (tivisse) [inline addons/hr_recruitment/views/hr_recruitment_views.xml:886]: Useless diff

File: addons/hr_recruitment_survey/views/hr_job_views.xml [hunk 14]
```diff
@@ -1,5 +1,11 @@
 <?xml version="1.0" encoding="utf-8"?>
 <odoo>
+    <record id="test_survey_view" model="ir.actions.act_window">
+        <field name="name">Survey</field>
```
Comment (tivisse) [inline addons/hr_recruitment_survey/views/hr_job_views.xml:4]: Please follow the guidelines.

https://www.odoo.com/documentation/13.0/reference/guidelines.html#xml-ids-and-naming

File: addons/mail/static/src/js/basic_view.js [hunk 15]
```diff
@@ -17,7 +17,8 @@ BasicView.include({
             hasRecordReloadOnAttachmentsChanged: post_refresh === 'always',
             hasRecordReloadOnMessagePosted: !!post_refresh,
             hasRecordReloadOnFollowersUpdate: !!followers_post_refresh,
-            isActivityBoxVisible: this._getFieldOption('message_ids', 'open_attachments', false),
+            isActivityBoxVisible: this._getFieldOption('message_ids', 'open_attachments', false)
```
Comment (tivisse) [inline addons/mail/static/src/js/basic_view.js:20]: 1/ Why ? 
2/ Should be done in a separate commit, in which we explain why, so that I don't have to ask it :D

File: addons/website_hr_recruitment/models/hr_recruitment.py [hunk 16]
```diff
@@ -10,19 +10,30 @@
...
         for source in self:
-            source.url = urls.url_join(base_url, "%s?%s" % (source.job_id.website_url,
-                urls.url_encode({
-                    'utm_campaign': self.env.ref('hr_recruitment.utm_campaign_job').name,
-                    'utm_medium': self.env.ref('utm.utm_medium_website').name,
-                    'utm_source': source.source_id.name
-                })
-            ))
+            link_tracker = self.env['link.tracker'].create({
+                'url': str(web_base_url) + str(source.job_id.website_url),
```
Comment (tivisse) [inline addons/website_hr_recruitment/models/hr_recruitment.py:22]:
```suggestion
                'url': "%s%s" % (web_base_url + source.job_id.website_url)
```

You sure it's that one of the elements is never False ?

File: addons/website_hr_recruitment/models/hr_recruitment.py [hunk 17]
```diff
@@ -10,19 +10,30 @@
...
+                'title': source.job_id.name,
+                'campaign_id': source.campaign_id.id,
+                'source_id': source.source_id.id,
+                'medium_id': self.env.ref('utm.utm_medium_website').id,
```
Comment (tivisse) [inline addons/website_hr_recruitment/models/hr_recruitment.py:26]: Check if it still exist. With raise_if_not_found.

File: addons/website_hr_recruitment/models/hr_recruitment.py [hunk 18]
```diff
@@ -10,19 +10,30 @@
...
+        web_base_url = self.env['ir.config_parameter'].sudo().get_param('web.base.url')
         for source in self:
-            source.url = urls.url_join(base_url, "%s?%s" % (source.job_id.website_url,
-                urls.url_encode({
-                    'utm_campaign': self.env.ref('hr_recruitment.utm_campaign_job').name,
-                    'utm_medium': self.env.ref('utm.utm_medium_website').name,
-                    'utm_source': source.source_id.name
-                })
-            ))
+            link_tracker = self.env['link.tracker'].create({
```
Comment (tivisse) [inline addons/website_hr_recruitment/models/hr_recruitment.py:21]: It's a very bad practice to create records in a compute method.

Don't do it that way.

But we can have a computed field that seek the link.tracked linked to the url.
As:
If someone clicks on the URL it creates a tracker. The compute method is able to find it, and the field would be empty if nobody clicked (it allow to avoid bloating the db for spam). Because at the end, we want to see the clicks, not the tracker which is a technical object.

File: addons/website_hr_recruitment/models/hr_recruitment.py [hunk 19]
```diff
@@ -10,19 +10,30 @@
 class RecruitmentSource(models.Model):
     _inherit = 'hr.recruitment.source'
 
-    url = fields.Char(compute='_compute_url', string='Url Parameters')
+    url = fields.Char(related='link_tracker_id.short_url', string='Url Parameters')
+    link_tracker_id = fields.Many2one('link.tracker', compute='_compute_link_tracker_id', store=True)
```
Comment (tivisse) [inline addons/website_hr_recruitment/models/hr_recruitment.py:14]: Would need a migration script to initialize the column + the value.

File: addons/website_hr_recruitment/models/hr_recruitment.py [hunk 20]
```diff
@@ -10,19 +10,30 @@
...
+                'medium_id': self.env.ref('utm.utm_medium_website').id,
+            })
+            source.link_tracker_id = link_tracker
+
+    def get_tracker_url(self):
+        self.ensure_one()
+        return {
+            'type': 'ir.actions.act_url',
+            'target': 'new',
+            'url': "%s+" % self.url
```
Comment (tivisse) [inline addons/website_hr_recruitment/models/hr_recruitment.py:35]: Why "+" ?

To remove IMO.

File: addons/website_hr_recruitment/models/hr_recruitment.py [hunk 21]
```diff
@@ -64,3 +75,39 @@ def set_open(self):
...
+            utm_source = self.env.ref("utm.utm_source_twitter")
+        elif network == 'linkedin':
+            utm_source = self.env.ref("utm.utm_source_linkedin")
+
+        sources = self.env['hr.recruitment.source'].search([
+            ("source_id", "=", utm_source.id),
+            ("job_id", "=", self.id)], order="create_date")
+
+        if sources:
+            source = sources[-1]
```
Comment (tivisse) [inline addons/website_hr_recruitment/models/hr_recruitment.py:94]: Use limit=1 in your search.

File: addons/website_hr_recruitment/models/hr_recruitment.py [hunk 22]
```diff
@@ -64,3 +75,39 @@ def set_open(self):
...
+            source_vals = {
+                'source_id': utm_source.id,
+                'job_id': self.id,
+            }
+            source = self.env['hr.recruitment.source'].create(source_vals)
+
+        if network == 'facebook':
+            url = 'https://www.facebook.com/sharer/sharer.php?u=%s' % source.url
+        elif network == 'twitter':
+            url = 'https://twitter.com/intent/tweet?tw_p=tweetbutton&text=Amazing job offer for %s! Check it live: %s' % (self.name, source.url)
```
Comment (tivisse) [inline addons/website_hr_recruitment/models/hr_recruitment.py:105]: Use url_encode. I would be surprised if it works.

File: addons/website_hr_recruitment/views/hr_job_views.xml [hunk 23]
```diff
@@ -14,8 +14,32 @@
...
-                    <a t-attf-href="#{record.website_url.raw_value}">Job Description</a>
+                    <a t-attf-href="#{record.website_url.raw_value}" class="mr-2">Job Description</a>
+                    <field name="is_published" widget='boolean_toggle'/>
                 </span>
+            </xpath> 
+        </field>
+    </record>
+
+    <record id="view_hr_job_kanban_referal_extends" model="ir.ui.view"> 
+        <field name="model">hr.job</field>
```
Comment (tivisse) [inline addons/website_hr_recruitment/views/hr_job_views.xml:25]: missing view name. 

+ follow the guidelines please.

**Review by tivisse — CHANGES_REQUESTED** [review]
Hi ! 

I've posted a first review on the PR. 

There are a lot a cosmetic points, but unfortunately I don't agree with the model modifications.

A job (an utm.source) can't be linked to a campaign. It makes no sense. I agree to introduce a recruitment campaign mechanism, but no by generating trackers on the fly on a compute method. 

Obviously, the task won't be merged in 14.0 in its current state. We should re-use the existing mechanism developed in the Social application to:
- Track links
- Track marketing campaigns

Linkedin / Twitter / Facebook are already managed, so everything to ready to automatize the post creation on social networks, the link creation. ...

It would be much better to reuse those integration features instead of a totally custom brol that don't really work as we think it should. The main issue is not really coming from this development (It's just the drop that breaks the camel's back
), the current version which is already mixing aliases and utms is already problematic.

Putting this in spec state to discuss with the Marketing Team about how we could do this properly, but the main specification is : Remove all the crap in hr_recruitment and create a hr_recruitment_social module. ;)

@FrancoisGe Sorry about that but I'm convinced that the final result will be much better (You would be able to post a job offer on several social medias at the same time for example) :)

Thanks for your work !

File: addons/hr_recruitment/models/hr_employee.py [hunk 24]
```diff
@@ -20,3 +22,18 @@ def _search_newly_hired_employee(self, operator, value):
...
+    def create(self, vals):
+        new_employee = super(HrEmployee, self).create(vals)
+        if new_employee.applicant_id:
+            new_employee.applicant_id.message_post_with_view(
+                        'hr_recruitment.applicant_hired_template',
+                        values={'applicant': new_employee.applicant_id},
+                        subtype_id=self.env.ref("hr_recruitment.mt_applicant_hired").id)
+        if new_employee.job_id:
+            new_employee.job_id.write({'no_of_hired_employee': new_employee.job_id.no_of_hired_employee + 1})
+            new_employee.job_id.message_post(
```
[the remark this answers is not in the document]
Comment (FrancoisGe) [reply addons/hr_recruitment/models/hr_employee.py:36]: Yes, it's a decision of the po

[the remark this answers is not in the document]
Comment (tivisse) [reply addons/hr_recruitment/models/hr_employee.py:36]: As the messages won't be visible on the interface, let's remove this ;)
